src/KubernetesClient/Informers/ResourceObservableExtensions.cs

```diff
@@ -0,0 +1,294 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reactive.Concurrency;
+using System.Reactive.Disposables;
+using System.Reactive.Linq;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Dataflow;
+using k8s.Informers.Cache;
+using k8s.Informers.Notifications;
+using Microsoft.Extensions.Logging;
+
+namespace k8s.Informers
+{
+    public static class ResourceObservableExtensions
+    {
+        /// <summary>
+        ///     Connects source block which publishes list of <see cref="ResourceEvent{TResource}" /> to action block
+        ///     which invokes processing function specified by <paramref name="action" /> for each received item.
+        /// </summary>
+        /// <param name="workerQueue">The source action block to attach to</param>
+        /// <param name="action">The action to invoke for each received batch of <see cref="ResourceEvent{TResource}" /></param>
+        /// <param name="parallelWorkers">Number of allowed parallel invocations of <paramref name="action" />. Default is 1</param>
+        /// <typeparam name="TResource">The resource type</typeparam>
+        /// <returns>The disposable that disconnects from the <paramref name="workerQueue" /> when disposed of</returns>
+        public static IDisposable ProcessWith<TResource>(this ISourceBlock<List<ResourceEvent<TResource>>> workerQueue, Func<List<ResourceEvent<TResource>>, Task> action, ILogger logger, int parallelWorkers = 1)
+        {
+            var actionBlock = new ActionBlock<List<ResourceEvent<TResource>>>(action, new ExecutionDataflowBlockOptions
+            {
+                BoundedCapacity = parallelWorkers, // don't buffer more messages then we are actually able to work on
+                MaxDegreeOfParallelism = parallelWorkers
+            });
+            actionBlock.Completion.ContinueWith(x =>
+            {
+                if (x.IsFaulted) logger.LogCritical(x.Exception.Flatten(), "Controller encountered a critical error");
+            });
+            return workerQueue.LinkTo(actionBlock, new DataflowLinkOptions {PropagateCompletion = true});
+        }
+
+        public static IObservable<ResourceEvent<TResource>> WithResets<TResource>(this IObservable<ResourceEvent<TResource>> source, Func<List<ResourceEvent<TResource>>, IEnumerable<ResourceEvent<TResource>>> action)
+        {
+            return source.WithResets(action, e => e);
+        }
+
+        public static IObservable<TOut> WithResets<TResource, TOut>(this IObservable<ResourceEvent<TResource>> source, Func<List<ResourceEvent<TResource>>, IEnumerable<ResourceEvent<TResource>>> resetSelector, Func<ResourceEvent<TResource>, TOut> itemSelector)
+        {
+            var resetBuffer = new List<ResourceEvent<TResource>>();
```

Comment:
`resetBuffer` will be shared across all subscriptions and I do not believe that is correct.

The declaration should be moved into the lambda supplied to `Create`.